hnn_core/extracellular.py

```diff
@@ -0,0 +1,587 @@
+            for jj in range(len(voltages[ii])):
+                if len(times) != len(voltages[ii][jj]):
+                    raise ValueError('length of times and voltages must match,'
+                                     f' got {len(times)} and '
+                                     f'{len(voltages[ii][jj])} for trial{ii}, '
+                                     f'channel {jj}')
+
+        self.positions = positions
+        self.n_contacts = len(self.positions)
+        self.conductivity = conductivity
+        self.method = method
+        self.min_distance = min_distance
+
+        self._times = times
+        self._data = voltages
+
+    def __getitem__(self, trial_no):
+        try:
+            if isinstance(trial_no, int):
+                return_data = [self._data[trial_no]]
+            elif isinstance(trial_no, slice):
+                return_data = self._data[trial_no]
+            elif isinstance(trial_no, (list, tuple)):
+                return_data = [self._data[trial] for trial in trial_no]
+            else:
+                raise TypeError(f'trial index must be int, slice or list-like,'
+                                f' got: {trial_no} which is {type(trial_no)}')
+        except IndexError:
+            raise IndexError(f'the data contain {len(self)} trials, the '
+                             f'indices provided are out of range: {trial_no}')
+        return ExtracellularArray(self.positions,
+                                  conductivity=self.conductivity,
+                                  method=self.method,
+                                  times=self.times,
+                                  voltages=return_data)
+
+    def __repr__(self):
+        class_name = self.__class__.__name__
+        msg = (f'{self.n_contacts} electrodes, '
+               f'conductivity={self.conductivity}, method={self.method}')
+        if len(self._data) > 0:
+            msg += f' | {len(self._data)} trials, {len(self.times)} times'
+        else:
+            msg += ' (no data recorded yet)'
+        return f'<{class_name} | {msg}>'
+
+    def __len__(self):
+        return len(self._data)  # length == number of trials
+
+    def copy(self):
+        """Return a copy of the ExtracellularArray instance
+
+        Returns
+        -------
+        array_copy : instance of ExtracellularArray
+            A copy of the array instance.
+        """
+        return deepcopy(self)
+
+    @property
+    def times(self):
+        return self._times
+
+    @property
+    def voltages(self):
+        return self._data
+
+    @property
+    def sfreq(self):
+        """Return the sampling rate of the extracellular data."""
+        if len(self.times) == 0:
+            return None
+        elif len(self.times) == 1:
+            raise RuntimeError('Sampling rate is not defined for one sample')
+
+        dT = np.diff(self.times)
+        Tsamp = np.median(dT)
+        if np.abs(dT.max() - Tsamp) > 1e-3 or np.abs(dT.min() - Tsamp) > 1e-3:
```

Comment:
glad to see I'm not the only one who doesn't fully trust the "fixed" time step ODE solver 😄

Reply:
Potentially devastating and definitely pathological!